### PR TITLE
Route MoonPay iOS buys through a Private Relay interstitial check

### PR DESCRIPTION
**Pull request by j0ntz**
### Technical Design Document

[moonpay-private-relay-interstitial.md](https://github.com/EdgeApp/edge-react-gui/blob/jon/moonpay-relay-interstitial/src/docs/moonpay-private-relay-interstitial.md)

### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

[EdgeApp/edge-info-server#160](https://github.com/EdgeApp/edge-info-server/pull/160) (soft dependency: without it the app falls back to today's bound flow, so this PR is safe to land first)

### Requirements

If you have made **any** visual changes to the GUI. Make sure you have:

- [ ] Tested on iOS device
- [ ] Tested on Android device
- [ ] Tested on small-screen device (iPod Touch)
- [ ] Tested on large-screen device (tablet)
- [ ] iCloud+ Private Relay on-device buy pass (MoonPay enforcement for our key stays off until this completes)

No visual changes: only the URL handed to the external Safari view differs; the relay pass above is the outstanding device item.

### Description

[Asana task](https://app.asana.com/0/0/1217471281005732/f)

iOS buys open the MoonPay widget through the server's Private Relay relay-check interstitial instead of a directly IP-bound URL, so the server can compare the app-fetch and Safari-view egress addresses and sign the widget URL with or without the binding. Any interstitial failure falls back to today's bound flow, so buys cannot regress; sell and Android are untouched. `MOONPAY_RELAY_CHECK_SIGN_PROXY` (dev builds only) makes the divergence branch reproducible on a simulator.

The TDD linked above carries the full design for both repos. The physical-device Private Relay pass is the outstanding manual item; MoonPay enforcement stays off until it completes.

---

> [!NOTE]
> **Medium Risk**
> Changes the critical path for iOS fiat on-ramp URL signing and IP binding, with intentional fallback to the old flow on failure; paired info-server behavior determines whether relay users succeed under MoonPay enforcement.
> 
> **Overview**
> **iOS MoonPay buys** no longer open a directly IP-bound widget URL in Safari. They first request a **relay-check interstitial** from the info server (`relayCheck: true` on `signUrl`); that URL is opened in `SFSafariViewController` so the server can compare app-fetch vs Safari egress and redirect to a signed widget **with or without** `allowedIpAddress`. **Sell and Android buy** still use direct bound signing.
> 
> If the interstitial request fails (network, old server, empty URL), the buy path **falls back** to today’s bound `signMoonpayUrl` flow so buys should not regress without the paired info-server deploy.
> 
> `moonpaySign.ts` centralizes `postSignUrl` (shared timeout and error handling). **Dev-only** `MOONPAY_RELAY_CHECK_SIGN_PROXY` reroutes the relay-check POST through an alternate egress for simulator testing of the unbound branch.
> 
> Adds unit tests for signing and interstitial fetch, a CHANGELOG entry, and a cross-repo TDD in `moonpay-private-relay-interstitial.md`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2b5d37c24e338979c7cfda14338746b12e6d0191. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by j0ntz** [comment]
## 📸 Test evidence

**moonpay buy quotes**

moonpay buy quotes

**safari widget fallback bound**

safari widget fallback bound

**safari interstitial unbound**

safari interstitial unbound

_Captured by the agent's in-app test run (build-and-test)._

File: src/plugins/gui/providers/moonpaySign.ts [hunk 1]
```diff
@@ -1,28 +1,81 @@
...
+  const signProxy = __DEV__ ? ENV.MOONPAY_RELAY_CHECK_SIGN_PROXY : undefined
+  const reply = await postSignUrl({ url, relayCheck: true }, signProxy)
+  const { interstitialUrl } = asMoonpayRelayCheckResponse(reply)
+  // An empty URL would open a blank Safari view with no error; throwing here
+  // routes the caller onto its bound-URL fallback instead.
+  if (interstitialUrl === '') {
+    throw new Error('Moonpay relay check returned an empty interstitial URL')
+  }
+  return interstitialUrl
+}
```
Comment (paullinator) [inline src/plugins/gui/providers/moonpaySign.ts:92]: **Warning:** No new unit tests cover fetchMoonpayInterstitialUrl, postSignUrl's MOONPAY_RELAY_CHECK_SIGN_PROXY / fetchWaterfall branch, the empty-interstitialUrl throw, or the iOS Platform.OS fallback in moonpayRampPlugin. The TDD (§8) only lists server mocha tests and Maestro/manual device passes for the GUI side, so the new client helpers ship without automated regression coverage despite being easily mockable (fetchInfo/fetchWaterfall).

**Recommendation:** Add Jest tests: mock fetchInfo/fetchWaterfall to assert relayCheck:true body + interstitialUrl parsing, empty-URL throw, non-OK status, and that a non-empty signProxy calls fetchWaterfall on the trimmed base. Optionally extract the iOS-vs-Android URL resolution so the Platform branch and .catch → signMoonpayUrl fallback can be asserted without mounting the full ramp plugin.

File: src/plugins/gui/providers/moonpaySign.ts [hunk 2]
```diff
@@ -1,28 +1,81 @@
...
+          [signProxy.replace(/\/+$/, '')],
+          SIGN_URL_PATH,
+          options,
+          SIGN_URL_TIMEOUT_MS
+        )
+      : await fetchInfo(SIGN_URL_PATH, options, SIGN_URL_TIMEOUT_MS)
+  if (!response.ok) {
+    throw new Error(`Moonpay URL signing failed: ${response.status}`)
+  }
+  return await response.json()
```
Comment (paullinator) [inline src/plugins/gui/providers/moonpaySign.ts:50]: **Suggestion:** postSignUrl is a good shared helper, but body is typed as object and HTTP failures always throw "Moonpay URL signing failed", including relayCheck requests. That weakens call-site type safety and makes interstitial fallback logs misleading.

**Recommendation:** Type the body as { url: string; relayCheck?: boolean } and use a mode-aware error message.

File: src/plugins/ramps/moonpay/moonpayRampPlugin.ts [hunk 3]
```diff
@@ -838,10 +842,33 @@ export const moonpayRampPlugin: RampPluginFactory = (
...
+                    )
+                    return await signMoonpayUrl(urlObj.href)
+                  })
+                } else {
+                  webViewUrl = await signMoonpayUrl(urlObj.href)
+                }
 
                 deeplinkToken = await openExternalWebView({
-                  url: signedUrl,
+                  url: webViewUrl,
```
Comment (paullinator) [inline src/plugins/ramps/moonpay/moonpayRampPlugin.ts]: **Suggestion:** webViewUrl is opened via openExternalWebView (SafariView / Custom Tabs), not a WebView; after the interstitial change the name no longer describes the value well.

**Recommendation:** Rename to openUrl or widgetUrl for clarity at the openExternalWebView call site.

File: src/plugins/ramps/moonpay/moonpayRampPlugin.ts [hunk 4]
```diff
@@ -838,10 +842,33 @@ export const moonpayRampPlugin: RampPluginFactory = (
...
+                if (Platform.OS === 'ios') {
+                  webViewUrl = await fetchMoonpayInterstitialUrl(
+                    urlObj.href
+                  ).catch(async (error: unknown) => {
+                    console.log(
+                      'Moonpay relay check unavailable, using bound URL: ' +
+                        String(error)
+                    )
+                    return await signMoonpayUrl(urlObj.href)
+                  })
```
Comment (paullinator) [inline src/plugins/ramps/moonpay/moonpayRampPlugin.ts:866]: **Suggestion:** The .catch handler is marked async only to return await signMoonpayUrl(...); the extra async/await is unnecessary.

**Recommendation:** Use .catch((error: unknown) => { console.log(...); return signMoonpayUrl(urlObj.href) }).

File: src/docs/moonpay-private-relay-interstitial.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+}
+```
+
+The Safari view opens `webViewUrl` exactly as it opened the signed URL before; the 302 delivers the user to the same widget, and the existing deeplink return path (`redirectURL` to `deep.edge.app`) is unchanged. A server that does not yet support relay-check mode, a network failure, or a malformed response all land in the catch and produce today's bound flow, so the app change cannot regress buys even against un-upgraded servers.
+
+One consequence of the fallback: when the relay-check POST fails and the plain retry succeeds against an upgraded server, a relaying v2 user is indistinguishable from the live app and lands on the update page. That session was already degraded (the primary path failed), and the page's advice is merely unnecessary rather than wrong.
+
+The dev flag: `ENV.MOONPAY_RELAY_CHECK_SIGN_PROXY` (declared in `src/envConfig.ts`, empty by default) reroutes this one POST through an alternate egress, so the server observes divergent A1/A2 on a simulator and the unbound branch is reproducible. The flag is read behind `__DEV__`, so release builds cannot reach it ([decision 3](#103-divergence-testing-flag-lives-client-side)).
```
Comment (paullinator) [inline src/docs/moonpay-private-relay-interstitial.md]: **Suggestion:** TDD says ENV.MOONPAY_RELAY_CHECK_SIGN_PROXY is "empty by default", but envConfig declares it as asOptional(asString) with no default, so an absent key is undefined (not "").

**Recommendation:** Rephrase to "absent/undefined by default (and ignored when empty)" to match asOptional(asString) and the signProxy guard.

File: src/plugins/ramps/moonpay/moonpayRampPlugin.ts [hunk 6]
```diff
@@ -838,10 +842,33 @@ export const moonpayRampPlugin: RampPluginFactory = (
...
+
+                // On iOS the widget opens in an SFSafariViewController whose
+                // traffic can egress through iCloud Private Relay, so a URL
+                // bound to the app-fetch IP is guaranteed to mismatch what
+                // MoonPay observes for relay users. Open the server's
+                // relay-check interstitial instead: it observes the Safari
+                // view's own egress and 302s to the widget URL signed with or
+                // without the IP binding. Any interstitial failure falls back
+                // to today's bound flow. Android (Custom Tabs, app network
+                // stack) keeps the fully bound flow.
```
Comment (paullinator) [inline src/plugins/ramps/moonpay/moonpayRampPlugin.ts]: **Suggestion:** Inline comment calls Android Custom Tabs the "app network stack". Sell's guiPluginWebView is the app stack; Custom Tabs use the device/browser stack (same egress as app fetch, but not the app WebView stack).

**Recommendation:** Align with the TDD: e.g. "Android (Custom Tabs, device network stack, same egress as app fetch) keeps the fully bound flow."

File: src/plugins/gui/providers/moonpaySign.ts [hunk 7]
```diff
@@ -1,28 +1,81 @@
...
+}
+
 /**
  * Ask the info server to bind a Moonpay widget URL to the caller's public IP
  * and sign it. Moonpay's on-ramp security upgrade refuses to load widget URLs
  * that are not signed and IP-bound, so every buy/sell widget URL must be routed
  * through here before it is opened.
```
Comment (paullinator) [inline src/plugins/gui/providers/moonpaySign.ts]: **Suggestion:** signMoonpayUrl's JSDoc still says every buy/sell widget URL must be routed through that function before opening. After this PR, the iOS buy happy path uses fetchMoonpayInterstitialUrl instead.

**Recommendation:** Update the JSDoc to say direct IP-bound signing (sell, Android buy, and iOS buy fallback); point iOS buy's primary path at fetchMoonpayInterstitialUrl.

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 ## Unreleased (develop)
 
+- added: MoonPay iOS buys open through a Private Relay interstitial check, so IP-match enforcement cannot lock out iCloud Private Relay users; any interstitial failure falls back to the existing bound flow.
```
Comment (paullinator) [inline CHANGELOG.md]: **Suggestion:** CHANGELOG uses "added:" for a change to the existing MoonPay IP-bound buy handoff. Nearby history already has "changed:" for signing/binding.

**Recommendation:** Prefer "changed:" (keep the same explanation about interstitial + fallback).

**Review by paullinator — COMMENTED** [review]
## Additional Findings

- **suggestion**: PR Requirements leaves all device-test boxes unchecked with "No visual changes", while the Description/TDD correctly call out that the physical-device Private Relay pass is still outstanding and that MoonPay enforcement must stay off until it completes.
  - Keep the no-visual-changes note, but add an explicit checklist item for the iCloud+ Private Relay on-device buy pass, and leave it unchecked until done.

File: src/plugins/gui/providers/moonpaySign.ts (hunk 1 above)
Comment (j0ntz) [reply src/plugins/gui/providers/moonpaySign.ts:92]: Fixed in 2b5d37c: new src/__tests__/plugins/gui/providers/moonpaySign.test.ts mocks fetchInfo/fetchWaterfall and covers the relayCheck:true body and interstitialUrl parsing, the empty-URL throw, non-OK errors for both modes, and the trimmed sign-proxy routing through fetchWaterfall with the 10s timeout (6 tests).

File: src/plugins/gui/providers/moonpaySign.ts (hunk 2 above)
Comment (j0ntz) [reply src/plugins/gui/providers/moonpaySign.ts:50]: Fixed in 2b5d37c: postSignUrl takes { url: string; relayCheck?: boolean } and the non-OK error is mode-aware (Moonpay relay check failed vs Moonpay URL signing failed).

File: src/plugins/ramps/moonpay/moonpayRampPlugin.ts (hunk 3 above)
Comment (j0ntz) [reply src/plugins/ramps/moonpay/moonpayRampPlugin.ts]: Fixed in 2b5d37c: renamed to openUrl. On the iOS primary path the value is the interstitial rather than the widget URL, so openUrl (the URL handed to openExternalWebView) stays accurate in both cases.

File: src/plugins/ramps/moonpay/moonpayRampPlugin.ts (hunk 4 above)
Comment (j0ntz) [reply src/plugins/ramps/moonpay/moonpayRampPlugin.ts:866]: Fixed in 2b5d37c: the catch returns signMoonpayUrl(urlObj.href) directly, no async/await wrapper.

File: src/docs/moonpay-private-relay-interstitial.md (hunk 5 above)
Comment (j0ntz) [reply src/docs/moonpay-private-relay-interstitial.md]: Fixed in 2b5d37c: the TDD now reads absent/undefined by default (and ignored when empty), matching asOptional(asString) and the signProxy guard.

File: src/plugins/ramps/moonpay/moonpayRampPlugin.ts (hunk 6 above)
Comment (j0ntz) [reply src/plugins/ramps/moonpay/moonpayRampPlugin.ts]: Fixed in 2b5d37c: the comment now says Android (Custom Tabs, device network stack, same egress as the app's fetch) keeps the fully bound flow, aligned with the TDD.

File: src/plugins/gui/providers/moonpaySign.ts (hunk 7 above)
Comment (j0ntz) [reply src/plugins/gui/providers/moonpaySign.ts]: Fixed in 2b5d37c: the JSDoc names direct bound signing as the sell / Android buy / iOS-fallback path and points the iOS buy primary path at fetchMoonpayInterstitialUrl.

File: CHANGELOG.md (hunk 8 above)
Comment (j0ntz) [reply CHANGELOG.md]: Fixed in 2b5d37c: the entry moved to the changed: group with the same explanation.

**Comment by j0ntz** [comment]
Top-level review note ([review 4965479992](https://github.com/EdgeApp/edge-react-gui/pull/6151#pullrequestreview-4965479992)) addressed: the Requirements checklist now carries an explicit unchecked item for the iCloud+ Private Relay on-device buy pass (MoonPay enforcement stays off until it completes), alongside the no-visual-changes note.

**Comment by j0ntz** [comment]
Addressed (Requirements now carries the unchecked iCloud+ Private Relay on-device buy pass item; details in the earlier reply comment).
